cocotb/drivers/amba.py

```diff
@@ -220,98 +221,176 @@ class AXI4Slave(BusDriver):
+            delay > 0: wait 'delay' clock cycles after valid to assert ready.
+        """
+        ready <= (0, 1)[delay < 0]
+
+        #check if valid is already asserted
+        if not valid.value:
+            yield RisingEdge(valid)
+
+        for i in range(delay):
+            yield RisingEdge(clock)
+
+        ready <= 1
+        yield RisingEdge(clock)
+        ready <= 0
+
+    @cocotb.coroutine
+    def _write_address(self):
+        """Monitor the Write Address channel for transactions
+        """
+        yield self._saxi_handshake(self.clock,
+                                   self.bus.AWREADY,
+                                   self.bus.AWVALID,
+                                   self.awready_delay)
+
+        awaddr = int(self.bus.AWADDR)
+        awlen = int(self.bus.AWLEN)
+        awsize = int(self.bus.AWSIZE)
+        awburst = int(self.bus.AWBURST)
+        awprot = int(self.bus.AWPROT)
+        awid = int(self.bus.AWID)
+
+        return awaddr, awlen, awsize, awburst, awprot, awid
+
     @cocotb.coroutine
     def _write_data(self):
-        clock_re = RisingEdge(self.clock)
+        """Monitor the Write Address channel for transactions
+        """
+        last = 0
+        wdata = []
+        wstrb = []
+        while not last:
+            yield self._saxi_handshake(self.clock,
+                                       self.bus.WREADY,
+                                       self.bus.WVALID,
+                                       self.wready_delay)
+
+            wdata.append(int(self.bus.WDATA))
+            wstrb.append(int(self.bus.WSTRB))
+            last = int(self.bus.WLAST)
+            wid = int(self.bus.WID) #should be the same for each beat
+        return wdata, wid, wstrb
 
+    @cocotb.coroutine
+    def _write_transaction(self):
+        """Process write transactions
+
+        This requires both wdata and awaddr to complete.
+        TODO support out of order threads
+        """
         while True:
-            while True:
-                self.bus.WREADY <= 0
-                yield ReadOnly()
-                if self.bus.AWVALID.value:
-                    self.bus.WREADY <= 1
-                    break
-                yield clock_re
+            self.bus.BVALID <= 0
+            (
+                (wdata, wid, wstrb),
+                (awaddr, awlen, awsize, awburst, awprot, awid)
+            ) = yield Combine(
+                self._write_data(),
+                self._write_address()
+            )
+
+            self.log.info("Write Transaction:\n" +
+                "\t0x{:08X} <= 0x{:016X}\n".format(awaddr, wdata[0]) +
+                reduce((lambda d, x:d + "\t{:14}0x{:016X}\n".format('', x)), wdata[1:], ''))
+
+            bresp = self.do_slave_write(
+                awaddr, awlen, awsize, awburst, awprot, awid,
+                wdata, wid, wstrb)
+
+            for i in range(self.bresp_delay-1):
+                yield RisingEdge(self.clock)
```

Comment:
I'd argue this belongs within `do_slave_write`

Reply:
I was thinking that do_slave_write wouldn't be a coroutine, just a simple python function that could be overridden to allow custom slave behaviour.  In which case simulation timing would be left to _write_transaction.

Reply:
If you override it as a coroutine, then the simulation timing could depend on the memory address (eg, emulating an MMU with different memory latencies).

Remember that it's always possible to write a trivial coroutine that behaves like a function, but not vice versa.

Reply:
Good point.  Though bresp_delay would make less sense in its current form.